### PR TITLE
Collection spacer with component

### PR DESCRIPTION
**Pull request by brphelps**
### Summary
Sample implementation looking for very rough 👍 or 👎 on direction before I start polishing.

Syntax for usage is the `collection_spacer` on `with_collection`: 
```
    render_inline(ProductComponent.with_collection(products, notice: "On sale", collection_spacer: CollectionSpacerComponent))
```

### CODE REVIEW

File: lib/view_component/collection.rb [hunk 1]
```diff
@@ -10,19 +10,23 @@ def render_in(view_context, &block)
       @component.compile(raise_errors: true)
       @component.validate_collection_parameter!(validate_default: true)
 
-      @collection.map do |item|
+      @collection.each_with_index.map do |item, index|
```
Comment (BlakeWilliams) [inline lib/view_component/collection.rb:13]: I wonder if we can continue to use `.map` here, and rely on `iterator.first`? https://api.rubyonrails.org/classes/ActionView/PartialIteration.html

**Review by BlakeWilliams — COMMENTED** [review]
Added one small comment, but this approach looks good to me. 👍

File: lib/view_component/collection.rb (hunk 1 above)
Comment (brphelps) [reply lib/view_component/collection.rb:13]: That makes sense to me, I didn't know it existed :D

**Comment by brphelps** [comment]
@BlakeWilliams  -- I'm a little confused about the failing checks. Is it a specific code coverage number not being met?

**Comment by BlakeWilliams** [comment]
> @BlakeWilliams -- I'm a little confused about the failing checks. Is it a specific code coverage number not being met?

I think the output is misleading, it looks like if there's a failure code coverage is still being run, and checked. We should probably fix that. :)

The actual error looks like: `File does not exist: pry`

File: lib/view_component/collection.rb (hunk 1 above)
Comment (mellowfish) [reply lib/view_component/collection.rb:13]: You can also do `@collection.map.with_index do ...`

File: lib/view_component/base.rb [hunk 2]
```diff
@@ -173,8 +173,8 @@ class << self
       attr_accessor :source_location, :virtual_path
 
       # Render a component collection.
-      def with_collection(collection, **args)
-        Collection.new(self, collection, **args)
+      def with_collection(collection, collection_spacer: nil, **args)
+        Collection.new(self, collection, collection_spacer, **args)
```
Comment (joelhawksley) [inline lib/view_component/base.rb:177]: In keeping with our desire to match Rails APIs, how about we call this `spacer_component`? See https://guides.rubyonrails.org/layouts_and_rendering.html#spacer-templates.

```suggestion
      def with_collection(collection, spacer_component: nil, **args)
        Collection.new(self, collection, spacer_component, **args)
```

File: lib/view_component/collection.rb [hunk 3]
```diff
@@ -11,18 +11,22 @@ def render_in(view_context, &block)
...
         iterator.iterate!
-        content
+        pre_content != nil ? pre_content + content : content
       end.join.html_safe
     end
 
     private
 
-    def initialize(component, object, **options)
+    def initialize(component, object, collection_spacer, **options)
```
Comment (joelhawksley) [inline lib/view_component/collection.rb:25]:
```suggestion
    def initialize(component, object, spacer_component, **options)
```

File: lib/view_component/collection.rb [hunk 4]
```diff
@@ -11,18 +11,22 @@ def render_in(view_context, &block)
...
       @component = component
       @collection = collection_variable(object || [])
       @options = options
+      @collection_spacer = collection_spacer
```
Comment (joelhawksley) [inline lib/view_component/collection.rb:29]:
```suggestion
      @spacer_component = spacer_component
```

File: lib/view_component/collection.rb [hunk 5]
```diff
@@ -11,18 +11,22 @@ def render_in(view_context, &block)
       @component.validate_collection_parameter!(validate_default: true)
 
       @collection.map do |item|
+        if !iterator.first? && @collection_spacer != nil
+          pre_content = @collection_spacer.new(item: item, index: index).render_in(view_context, &block)
+        end
```
Comment (joelhawksley) [inline lib/view_component/collection.rb:16]:
```suggestion
        if !iterator.first? && @spacer_component
          spacer_content = @spacer_component.new(item: item, index: index).render_in(view_context, &block)
        end
```

File: lib/view_component/collection.rb [hunk 6]
```diff
@@ -11,18 +11,22 @@ def render_in(view_context, &block)
...
         content = @component.new(**component_options(item, iterator)).render_in(view_context, &block)
         iterator.iterate!
-        content
+        pre_content != nil ? pre_content + content : content
```
Comment (joelhawksley) [inline lib/view_component/collection.rb:19]:
```suggestion
        spacer_content != nil ? spacer_content + content : content
```

File: test/app/components/collection_spacer_component.html.erb [hunk 7]
```diff
@@ -0,0 +1,3 @@
+<div id="spacer-<%= @index %>">
+  <br/>Sample spacer... <br/>
+</div>
```
Comment (joelhawksley) [inline test/app/components/collection_spacer_component.html.erb:3]:
```suggestion
<div id="spacer-<%= @index %>">Sample spacer</div>

```

File: test/view_component/view_component_test.rb [hunk 8]
```diff
@@ -555,6 +555,13 @@ def test_render_collection
     assert_selector("p", text: "Mints counter: 2")
   end
 
+  def test_render_collection_with_spacer
+    products = [OpenStruct.new(name: "Radio clock"), OpenStruct.new(name: "Mints"), OpenStruct.new(name: "Toys")]
+    render_inline(ProductComponent.with_collection(products, notice: "On sale", collection_spacer: CollectionSpacerComponent))
+    assert_selector("#spacer-1", text: "Sample spacer...", count: 1)
+    assert_selector("#spacer-2", text: "Sample spacer...", count: 1)
```
Comment (joelhawksley) [inline test/view_component/view_component_test.rb:562]:
```suggestion
    assert_selector("#spacer-1", text: "Sample spacer", count: 1)
    assert_selector("#spacer-2", text: "Sample spacer", count: 1)
```

**Review by joelhawksley — APPROVED** [review]
Looks good @brphelps! Mind adding a CHANGELOG entry, your name to the bottom of `docs/index.md`, and some sort of documentation to `docs/index.md`?

File: test/view_component/view_component_test.rb [hunk 9]
```diff
@@ -555,6 +555,13 @@ def test_render_collection
     assert_selector("p", text: "Mints counter: 2")
   end
 
+  def test_render_collection_with_spacer
+    products = [OpenStruct.new(name: "Radio clock"), OpenStruct.new(name: "Mints"), OpenStruct.new(name: "Toys")]
+    render_inline(ProductComponent.with_collection(products, notice: "On sale", collection_spacer: CollectionSpacerComponent))
```
Comment (BlakeWilliams) [inline test/view_component/view_component_test.rb:560]:
```suggestion
    render_inline(ProductComponent.with_collection(products, notice: "On sale", spacer_component: CollectionSpacerComponent))
```

File: lib/view_component/collection.rb [hunk 10]
```diff
@@ -11,18 +11,22 @@ def render_in(view_context, &block)
       @component.validate_collection_parameter!(validate_default: true)
 
       @collection.map do |item|
+        if !iterator.first? && @spacer_component
+          spacer_content = @spacer_component.new(item: item, index: iterator.index).render_in(view_context, &block)
```
Comment (joelhawksley) [inline lib/view_component/collection.rb:15]: Will all spacer components need to accept an `item` parameter? IMO we should look at using `component_options` or something like it here.

Comment (brphelps) [reply lib/view_component/collection.rb:15]: Is there a way to do this without making parameters required? The component_options is close to what I'd want, but I don't want the addition of parameters to break existing component usages.

Comment (joelhawksley) [reply lib/view_component/collection.rb:15]: What do you mean by:

> I don't want the addition of parameters to break existing component usages

`component_options` should only pass in the options if they are present in the component method signature.

Comment (brphelps) [reply lib/view_component/collection.rb:15]: Sorry it took me so long to loop back on this -- Basically, my preference would be that there is a set of ambient info that a component could use -- maybe a given spacer doesn't care about index or item, for example, so it wouldn't need a parameter for those. I guess i don't understand how component options helps :D

**Comment by brphelps** [comment]
Closing this out, it's age old and I haven't touched dotcom UI in months and months D:

**Comment by joelhawksley** [comment]
@boardfish would you be interested in picking this one up? I feel like @brphelps got it pretty close.

**Comment by boardfish** [comment]
Sure thing. I don't know how much time I'll have to start on it before the end of the year, but I'll try and get to grips with the wider context.

I'll open a fresh PR based off the existing work so as not to ping @brphelps too regularly @brphelps, if this feature is something you'd potentially still make use of in working with ViewComponent, I'd appreciate it if you could drop a little more context about how you'd like it to work so I can do it full justice.

**Pull request closed without merging**